### PR TITLE
Feature/cdsct 275 retrieve encounter audits

### PR DESCRIPTION
**Pull request by cmarcus-nhs**
Depends on https://github.com/UECIT/cactus-ems/pull/49

### CODE REVIEW

File: EMS-UI/src/app/validation-report/validation-report.component.ts [hunk 1]
```diff
@@ -67,10 +70,28 @@ export class ValidationReportComponent implements OnInit {
...
         //TODO: handle errors properly
       ).catch(err => this.loadedSearchAudits = true);
   }
+
+  private groupByCase(interactions: Interaction[]) : Interaction[] {
+    let grouped = interactions.reduce(function(r, a) {
+      r[a.additionalProperties["caseId"]] = r[a.additionalProperties["caseId"]] || [];
+      r[a.additionalProperties["caseId"]].push(a);
+      return r;
+    }, Object.create(null));
```
Comment (mhouston-nhs) [inline EMS-UI/src/app/validation-report/validation-report.component.ts]: The intent behind this is probably better captured by a Map object, which seems to be the replacement for this idiom

File: EMS-UI/src/app/validation-report/validation-report.component.ts [hunk 2]
```diff
@@ -67,10 +70,28 @@ export class ValidationReportComponent implements OnInit {
...
+
+    var firstByCase = [];
+    let cases = Object.keys(grouped);
+    // Only need to display the first interaction
+    for (let i = 0; i < cases.length; i++) {
```
Comment (mhouston-nhs) [inline EMS-UI/src/app/validation-report/validation-report.component.ts]: Maps are also iterable, which may simplify this

File: src/main/java/uk/nhs/ctp/audit/sqs/AWSAuditSender.java [hunk 3]
```diff
@@ -55,6 +56,12 @@ public void sendAudit(AuditSession session) {
                 .withQueueUrl(loggingQueue)
                 .withMessageBody(mapper.writeValueAsString(session));
             sqsClient.sendMessage(request);
+        } catch (AmazonSQSException e) {
+            if (e.getStatusCode() == 413) {
+                log.warn("Audit request exceeded max size SQS can handle", e);
+                //TODO: CDSCT-338 - Should we be auditing/paging audit search audits or excluding things
```
Comment (mhouston-nhs) [inline src/main/java/uk/nhs/ctp/audit/sqs/AWSAuditSender.java:62]: The general advice with SQS is if the payload is too large, it should be stored in S3 and sent as a reference instead.

In this case, I think we probably want to exclude auditing other audit-related calls as it's outside the CDS API anyway

File: src/test/java/uk/nhs/ctp/auditFinder/AWSAuditFinderTest.java [hunk 4]
```diff
@@ -70,28 +64,22 @@
 
   @Test
   public void findAll_withNullClient_shouldFail() {
-    var authService = mock(TokenAuthenticationService.class);
-    when(authService.requireSupplierId()).thenReturn("test-supplier");
-    var auditFinder = new AWSAuditFinder(null, authService, mock(ObjectMapper.class));
+    var auditFinder = new AWSAuditFinder(null, authService, objectMapper);
 
     expectedException.expect(IllegalArgumentException.class);
```
Comment (mhouston-nhs) [inline src/test/java/uk/nhs/ctp/auditFinder/AWSAuditFinderTest.java:69]: Will this ever be the case? Specifying a null esClient seems unexpected, and presumably Spring injection will fail if there isn't one in the context?

File: src/test/java/uk/nhs/ctp/auditFinder/AWSAuditFinderTest.java [hunk 5]
```diff
@@ -162,6 +141,68 @@ public void findAll_withFailedParsing_shouldFail() throws IOException {
     auditFinder.findAll(76L);
   }
 
+  @Test
+  public void findEncounters_nullClient_shouldFail() {
+    var auditFinder = new AWSAuditFinder(null, authService, objectMapper);
+
+    expectedException.expect(IllegalArgumentException.class);
```
Comment (mhouston-nhs) [inline src/test/java/uk/nhs/ctp/auditFinder/AWSAuditFinderTest.java:148]: Same as the other null client test, this seems odd.
Also, the IllegalArgumentException thrown is wrong, as it's an argument of the constructor which was invalid, not this call.

File: EMS-UI/src/app/validation-report/validation-report.component.ts (hunk 1 above)
Comment (cmarcus-nhs) [reply EMS-UI/src/app/validation-report/validation-report.component.ts]: Updated this in latest commit, think it's much cleaner now.

File: src/main/java/uk/nhs/ctp/audit/sqs/AWSAuditSender.java (hunk 3 above)
Comment (cmarcus-nhs) [reply src/main/java/uk/nhs/ctp/audit/sqs/AWSAuditSender.java:62]: I'll add that to the ticket

File: src/test/java/uk/nhs/ctp/auditFinder/AWSAuditFinderTest.java (hunk 4 above)
Comment (cmarcus-nhs) [reply src/test/java/uk/nhs/ctp/auditFinder/AWSAuditFinderTest.java:69]: This is true, the ES client itself makes sure that a URL is specified. I'll remove these tests.

File: EMS-UI/src/app/validation-report/validation-report.component.ts (hunk 1 above)
Comment (mhouston-nhs) [reply EMS-UI/src/app/validation-report/validation-report.component.ts]: Agreed, I like the single-pass approach :)

**Pull request merged**